File: examples/python/dex_examples/patterns/workflow/intervention/manual_intervention_flow.py
# Copyright (c) 2022-2026 Super Durable, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Handles API-call failure scenarios with manual retry or skip intervention."""

from __future__ import annotations

from typing import Callable

from dex import (
    Attribute,
    Channel,
    Context,
    Flow,
    PersistenceSchema,
    Step,
    StepDecision,
    StepList,
    Wait,
    go_to,
    graceful_complete,
)


class Final(Step[None]):
    def __init__(self, number_of_retries: Attribute[int]) -> None:
        self.number_of_retries = number_of_retries

    def execute(self, context: Context, input: None) -> StepDecision:
        del input
        retries = self.number_of_retries.get(context)
        return graceful_complete(f"Workflow Completed. Number of retries: {retries}")


class GetData(Step[bool]):
    def __init__(
        self,
        error_provider: Callable[[], Error],
        final: Final,
        data_channel: Channel[str],
        number_of_retries: Attribute[int],
    ) -> None:
        self.error_provider = error_provider
        self.final = final
        self.data_channel = data_channel
        self.number_of_retries = number_of_retries

    def wait_for(self, context: Context, input: bool) -> Wait:
        del context, input
        print("Waiting for incoming data")
        return Wait.all_of(self.data_channel.for_one())

    def execute(self, context: Context, input: bool) -> StepDecision:
        if input:
            self.number_of_retries.set(context, self.number_of_retries.get(context) + 1)
        try:
            self._pretend_api_call(context)
        except ValueError:
            return go_to(self.error_provider(), None)
        return go_to(self.final, None)

    def _pretend_api_call(self, context: Context) -> None:
        results = self.data_channel.results(context)
        if results:
            data = results[0]
            print(f"Received data result: {data}")
            if data == "failed":
                raise ValueError("Non-retryable exception")


class Error(Step[None]):
    def __init__(
        self,
        get_data: GetData,
        final: Final,
        retry_signal: Channel[None],
        skip_signal: Channel[None],
    ) -> None:
        self.get_data = get_data
        self.final = final
        self.retry_signal = retry_signal
        self.skip_signal = skip_signal

    def wait_for(self, context: Context, input: None) -> Wait:
        del context, input
        return Wait.any_of(self.retry_signal.for_one(), self.skip_signal.for_one())

    def execute(self, context: Context, input: None) -> StepDecision:
        del input
        retry = bool(self.retry_signal.results(context))
        print(
            "signal received: "
            + (self.retry_signal.name if retry else self.skip_signal.name)
        )
        if retry:
            return go_to(self.get_data, True)
        return go_to(self.final, None)


class Init(Step[None]):
    def __init__(self, get_data: GetData, number_of_retries: Attribute[int]) -> None:
        self.get_data = get_data
        self.number_of_retries = number_of_retries

    def execute(self, context: Context, input: None) -> StepDecision:
        del input
        self.number_of_retries.set(context, 0)
        return go_to(self.get_data, False)


class ManualInterventionFlow(Flow[None]):
    INTERNAL_CHANNEL_COMMAND = "internal_channel_command"
    SIGNAL_CHANNEL_COMMAND_RETRY = "signal_channel_command_retry"
    SIGNAL_CHANNEL_COMMAND_SKIP = "signal_channel_command_skip"
    NUMBER_OF_RETRIES = "number_of_retries"

    data_channel = Channel(INTERNAL_CHANNEL_COMMAND, str)
    retry_signal = Channel[None](SIGNAL_CHANNEL_COMMAND_RETRY, type(None))
    skip_signal = Channel[None](SIGNAL_CHANNEL_COMMAND_SKIP, type(None))
    number_of_retries = Attribute(NUMBER_OF_RETRIES, int)

    def __init__(self) -> None:
        self.final = Final(self.number_of_retries)
        self.get_data = GetData(
            lambda: self.error,
            self.final,
            self.data_channel,
            self.number_of_retries,
        )
        self.error = Error(
            self.get_data,
            self.final,
            self.retry_signal,
            self.skip_signal,
        )
        self.init = Init(self.get_data, self.number_of_retries)

    def get_steps(self) -> StepList[None]:
        return StepList.start_step(self.init).other_steps(
            self.get_data,
            self.error,
            self.final,
        )

    def get_persistence_schema(self) -> PersistenceSchema:
        return PersistenceSchema.of(
            self.data_channel,
            self.retry_signal,
            self.skip_signal,
            self.number_of_retries,
        )

File: examples/java/src/main/java/io/superdurable/dex/patterns/intervention/ManualInterventionFlow.java
/*
 * Copyright (c) 2022-2026 Super Durable, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.superdurable.dex.patterns.intervention;

import io.superdurable.dex.Attribute;
import io.superdurable.dex.Channel;
import io.superdurable.dex.Context;
import io.superdurable.dex.Flow;
import io.superdurable.dex.PersistenceSchema;
import io.superdurable.dex.Step;
import io.superdurable.dex.StepDecision;
import io.superdurable.dex.StepList;
import io.superdurable.dex.Wait;
import org.springframework.stereotype.Component;

import java.util.List;

/**
 * Handles API-call failure scenarios with manual retry or skip intervention.
 */
@Component
public class ManualInterventionFlow implements Flow<Void> {
    public static final String INTERNAL_CHANNEL_COMMAND = "internal_channel_command";
    public static final String SIGNAL_CHANNEL_COMMAND_RETRY = "signal_channel_command_retry";
    public static final String SIGNAL_CHANNEL_COMMAND_SKIP = "signal_channel_command_skip";
    public static final String NUMBER_OF_RETRIES = "number_of_retries";

    public final Channel<String> dataChannel =
            Channel.define(INTERNAL_CHANNEL_COMMAND, String.class);
    public final Channel<Void> retrySignal =
            Channel.define(SIGNAL_CHANNEL_COMMAND_RETRY, Void.class);
    public final Channel<Void> skipSignal =
            Channel.define(SIGNAL_CHANNEL_COMMAND_SKIP, Void.class);
    public final Attribute<Integer> numberOfRetries =
            Attribute.define(NUMBER_OF_RETRIES, Integer.class);

    private final Init init = new Init();
    private final GetData getData = new GetData();
    private final Error error = new Error();
    private final Final finalStep = new Final();

    @Override
    public StepList<Void> getSteps() {
        return StepList.startStep(init).otherSteps(getData, error, finalStep);
    }

    @Override
    public PersistenceSchema getPersistenceSchema() {
        return PersistenceSchema.of(dataChannel, retrySignal, skipSignal, numberOfRetries);
    }

    final class Init implements Step<Void> {
        @Override
        public Class<Void> getInputType() {
            return Void.class;
        }

        @Override
        public StepDecision execute(final Context context, final Void input) {
            numberOfRetries.set(context, 0);
            return StepDecision.goTo(getData, false);
        }
    }

    final class GetData implements Step<Boolean> {
        @Override
        public Class<Boolean> getInputType() {
            return Boolean.class;
        }

        @Override
        public Wait waitFor(final Context context, final Boolean isRetry) {
            System.out.println("Waiting for incoming data");
            return Wait.until(dataChannel.forOne());
        }

        @Override
        public StepDecision execute(final Context context, final Boolean isRetry) {
            if (Boolean.TRUE.equals(isRetry)) {
                final Integer retries = numberOfRetries.get(context);
                numberOfRetries.set(context, retries + 1);
            }
            try {
                pretendApiCall(context);
            } catch (final Exception e) {
                return StepDecision.goTo(error, null);
            }
            return StepDecision.goTo(finalStep, null);
        }

        private void pretendApiCall(final Context context) {
            final List<String> results = dataChannel.getConditionResults(context);
            if (!results.isEmpty()) {
                final String data = results.get(0);
                System.out.println("Received data result: " + data);
                if ("failed".equals(data)) {
                    throw new IllegalArgumentException("Non-retryable exception");
                }
            }
        }
    }

    final class Error implements Step<Void> {
        @Override
        public Class<Void> getInputType() {
            return Void.class;
        }

        @Override
        public Wait waitFor(final Context context, final Void input) {
            return Wait.anyOf(retrySignal.forOne(), skipSignal.forOne());
        }

        @Override
        public StepDecision execute(final Context context, final Void input) {
            final boolean retry =
                    !retrySignal.getConditionResults(context).isEmpty();
            System.out.println("signal received: "
                    + (retry ? SIGNAL_CHANNEL_COMMAND_RETRY : SIGNAL_CHANNEL_COMMAND_SKIP));
            if (retry) {
                return StepDecision.goTo(getData, true);
            }
            return StepDecision.goTo(finalStep, null);
        }
    }

    final class Final implements Step<Void> {
        @Override
        public Class<Void> getInputType() {
            return Void.class;
        }

        @Override
        public StepDecision execute(final Context context, final Void input) {
            final Integer retries = numberOfRetries.get(context);
            return StepDecision.gracefulComplete(
                    "Workflow Completed. Number of retries: " + retries);
        }
    }
}
